src/contracts/SmartContract.sol

```diff
@@ -29,31 +31,42 @@ contract SmartContract {
     uint public extraTime;
     address public helperID;
     uint public initialTimestamp;
+    Status public contractStatus = Status.Setup;
 
     mapping(uint => PuzzlePart) public puzzleParts;
     uint amountOfPuzzleParts;
     uint public nextUnsolvedPuzzlePart = 0;
     address owner;
 
-    constructor(
+    constructor() public payable {
+        owner = msg.sender;
+    }
+
+    function initialize(
         uint[] memory _coins,
         uint _startTime,
         uint[] memory _extraTimes,
         uint[] memory _upperBounds,
         address _helperID
-    ) public payable {
+    ) public payable onlyOwner {
 
         require(_coins.length == _upperBounds.length, "The length of the coins and upperBounds arrays should be the same.");
         require(_coins.length > 0, "The length of the coins array should be greater than 0.");
         require(_helperID != address(0), "The helper ID should not be the zero address.");
+        require(contractStatus == Status.Setup, "Contract setup is already completed.");
 
-        startTime = _startTime;
-        helperID = _helperID;
-        initialTimestamp = block.timestamp;
+        if (amountOfPuzzleParts == 0) {
+            startTime = _startTime;
+            helperID = _helperID;
+            initialTimestamp = block.timestamp;
+            owner = msg.sender;
+        }
 
         uint receivedValue = msg.value;
+        uint currentIndex = amountOfPuzzleParts;
```

Comment:
does this need to be a separate variable?

Reply:
It does improve the readability of the code in my opinion, though can be removed

Reply:
I get your point, though `currentIndex + i` is a tad confusing

Reply:
Agree, changed it to `startIndex` to also be consistent with the `setCommitment`